### PR TITLE
Static backends

### PR DESCRIPTION
**Pull request by andrewlouis93**
**What this PR does / why we need it**:
* Want to be enable balance between backends defined by static upstream pools.
* As long as user implements logic to set `proxy_upstream_name` to the right pool, this module will handle the rest. 

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Special notes for your reviewer**:

### CODE REVIEW

File: rootfs/etc/nginx/lua/balancer/static_upstreams.lua [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+    return sb
+end
+
+-- If any static upstream matches this pattern, add to static_backends
+function _M.configure()
+    local upstreams = ngx_upstream.get_upstreams()
+    for _, upstream_name in ipairs(upstreams) do
+        if upstream_name ~= "upstream_balancer" then
+            local sb = create_static_backend(upstream_name)
+            static_backends[upstream_name] = sb
```
Comment (ElvinEfendi) [inline rootfs/etc/nginx/lua/balancer/static_upstreams.lua]: `sb` here has to be an instance of a load balancing implementation

**Comment by ElvinEfendi** [comment]
What about introducing `static_balancers` variable in addition to `balancers` so that we don't mock with that logic. Then in https://github.com/Shopify/ingress/pull/76/files#diff-b00d77a6df9c8c05a483044b08e6bc50R96 we just do
```
return balancers[backend_name] or static_balancers[backend_name]
```

File: rootfs/etc/nginx/lua/balancer/static_upstreams.lua (hunk 1 above)
Comment (ElvinEfendi) [reply rootfs/etc/nginx/lua/balancer/static_upstreams.lua]: so you can directly do something like
```
... = require("balancer.round_robin").new(sb)
```

File: rootfs/etc/nginx/lua/balancer/static_balancers.lua [hunk 2]
```diff
@@ -0,0 +1,83 @@
+local util = require('util')
+local split = require('util.split')
+local ngx_upstream = require("ngx.upstream")
+local implementations = require('balancer.implementations')
+
+local static_backends = {}
+local static_balancers = {}
+
+local DEFAULT_LB_ALG = "ewma"
```
Comment (andrewlouis93) [inline rootfs/etc/nginx/lua/balancer/static_balancers.lua]: @ElvinEfendi Should this be parameterized as an argument to `configure`?

File: rootfs/etc/nginx/lua/balancer/static_balancers.lua [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+
+            return endpoint, nil
+        end
+    end
+    return nil, "error in grabbing address & port" 
+end
+
+local function create_static_backend(upstream_name)
+    local sb = {}
+    sb.name = upstream_name
```
Comment (ElvinEfendi) [inline rootfs/etc/nginx/lua/balancer/static_balancers.lua]: can do
```
local sb = { name = upstream_name }
```

File: rootfs/etc/nginx/lua/balancer/static_balancers.lua [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+
+    sb.endpoints = ngx_upstream.get_servers(upstream_name)
+
+    for index, endpoint in ipairs(sb.endpoints) do
+        sb.endpoints[index] = marshal_endpoint(endpoint)
```
Comment (ElvinEfendi) [inline rootfs/etc/nginx/lua/balancer/static_balancers.lua]: I'd avoid updating the variable that you're iterating over. With the changes in my above comment you can do
```
local servers = ngx_upstream.get_servers(upstream_name)
for _, server in ipairs(servers) do
  table.insert(sb.endpoints, marshal_endpoint(server))
end
```

File: rootfs/etc/nginx/lua/balancer/static_balancers.lua [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+            endpoint.port = addr.port
+            endpoint.addr = nil
+
+            return endpoint, nil
+        end
+    end
+    return nil, "error in grabbing address & port" 
+end
+
+local function create_static_backend(upstream_name)
```
Comment (ElvinEfendi) [inline rootfs/etc/nginx/lua/balancer/static_balancers.lua]: the module name already sets the expectation that operations in this module are about static backends, therefore I'd just call it `create_backend`. If we would be generating something other than static then it'd be valuable to include that in the name (i.e `create_non_static_backend`).

File: rootfs/etc/nginx/lua/balancer/static_balancers.lua [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+    populate_static_balancers()
+end
+
+function _M.get()
+    return static_balancers
+end
+
+if _TEST then
+    _M.backends = function()
+        return static_backends
```
Comment (ElvinEfendi) [inline rootfs/etc/nginx/lua/balancer/static_balancers.lua]: why don't you use already exported `_M.get()` instead?

File: rootfs/etc/nginx/lua/balancer/static_balancers.lua [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+    end
+
+    _M.reset = function()
+        static_backends = {}
+        static_balancers = {}
+    end
```
Comment (ElvinEfendi) [inline rootfs/etc/nginx/lua/balancer/static_balancers.lua]: you could achieve this in the test module itself by
```
package.loaded["static_balancers"] = nil
static_balancers = require("static_balancers")
```

File: rootfs/etc/nginx/lua/balancer/static_balancers.lua (hunk 6 above)
Comment (andrewlouis93) [reply rootfs/etc/nginx/lua/balancer/static_balancers.lua]: `_M.get()` is for returning balancers, not backends

File: rootfs/etc/nginx/lua/balancer/static_balancers.lua [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+local implementations = require('balancer.implementations')
+
+local backends = {}
+local balancers = {}
+
+local DEFAULT_LB_ALG = "ewma"
+
+local _M = {}
+
+local function marshal_endpoint(endpoint)
```
Comment (ElvinEfendi) [inline rootfs/etc/nginx/lua/balancer/static_balancers.lua]: Maybe `parse_endpoint(server)`? When you say marshal I think of encoding an object into a text blob.

File: rootfs/etc/nginx/lua/balancer/static_balancers.lua [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+
+local backends = {}
+local balancers = {}
+
+local DEFAULT_LB_ALG = "ewma"
+
+local _M = {}
+
+local function parse_endpoint(endpoint)
+    if (not endpoint.address) or (not endpoint.port) then
```
Comment (ElvinEfendi) [inline rootfs/etc/nginx/lua/balancer/static_balancers.lua]: I don't understand this, what is it for? My understanding of the function is it receives a `server` and generates `endpoint` from it and returns that endpoint.

File: rootfs/etc/nginx/lua/balancer/static_balancers.lua [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+local implementations = require('balancer.implementations')
+
+local backends = {}
+local balancers = {}
+
+local DEFAULT_LB_ALG = "ewma"
+
+local _M = {}
+
+local function parse_endpoint(endpoint)
```
Comment (ElvinEfendi) [inline rootfs/etc/nginx/lua/balancer/static_balancers.lua]: maybe call the argument, `server`?

File: rootfs/etc/nginx/lua/balancer/static_balancers.lua [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+        end
+    end
+    return nil, "error in grabbing address & port"
+end
+
+local function create_backend(upstream_name)
+    local sb = {
+        name = upstream_name,
+        endpoints = {},
+        ['load-balance'] = DEFAULT_LB_ALG
```
Comment (ElvinEfendi) [inline rootfs/etc/nginx/lua/balancer/static_balancers.lua]: lets stick to `"` for strings.

and I liked your idea of specifying load balancing alg as an argument to `configure()`

Comment (andrewlouis93) [reply rootfs/etc/nginx/lua/balancer/static_balancers.lua]: > lets stick to " for strings.

Seems like is more natural to use double quotes for user visible strings and text, whatever appears in output or in files; and single quotes for option-like strings, passing to methods and generally in code. 

Can change if you feel strongly about this. 

> and I liked your idea of specifying load balancing alg as an argument to configure()

Sure, will do this in a subsequent PR.

File: rootfs/etc/nginx/lua/balancer/static_balancers.lua (hunk 9 above)
Comment (andrewlouis93) [reply rootfs/etc/nginx/lua/balancer/static_balancers.lua]: Are you asking about the condition on line 13? 

I think that's a fair comment. Seems like I wasn't trusting the abstraction https://github.com/openresty/lua-upstream-nginx-module#get_servers

**Comment by ElvinEfendi** [comment]
I think we have a fundamental issue with this approach. We get the list of servers only once in the beginning. This is not gonna work. We will have healthchecking module enabled for these static servers and if one of them goes down that module marks the server as down. But this balancer will still be picking that server.

Instead of `get_servers` I think we should be using `get_primary_peers`. And then iterate of the peers to filter out the ones that are down. And this should happen periodically (sync_backend?).

I'm sorry if my previous comments misled you.

--

Maybe we should step back and see if there's any other way of doing of what we wanna do.

**Comment by andrewlouis93** [comment]
We can certainly step back to discuss, I figure this approach would still enable us to move forwards, and that we could iterate on it on including the healthcheck module.

**Pull request closed without merging**